src/Common/src/TypeSystem/Interop/IL/MarshalUtils.cs

```diff
@@ -59,6 +59,10 @@ public static bool IsBlittableType(TypeDesc type)
         public static bool IsManagedSequentialType(TypeDesc type)
         {
             type = type.UnderlyingType;
+            if (type is MetadataType metadataType && metadataType.IsExplicitLayout)
```

Comment:
Is this doing the right think for mix of explicit layout and non-explicit layout fields?

Reply:
From my limited understanding the IsManagedSequential check in CoreCLR basically performs a series of AND operations on several boolean indicators when calculating the outcome. I believe this roughly corresponds to the current structure of the recursive method where basically all fast paths return false (when one of the boolean indicators is incorrect).

In this sense I believe that the presence of any instance fields with explicit layout types guarantees that the type will be identified as non-managed-sequential in the for() loop over the instance fields calling recursively into the IsManagedSequentialType check. I apologize if I misunderstood your concern; if that was the case, please clarify.

Reply:
What kind of testing have you done on this?

It may be useful to add temporary instrumentation that logs the value of IsManagedSequential in both crossgen and cpaot, run all tests we have around (coreclr, corefx, winforms, wpf, ...); and see whether it is 100% match between the two.

I am worried that the field layout plan we are on for cpaot is going to have never ending bug tail.

Reply:
Forgetting about `IsManagedSequential` for a moment and looking at the bigger picture - from my experience in messing around in the CoreCLR codebase, we'll need to 100% match all the p/invoke field marshallers to get the right answers for instance layout. This is because `IsBlittable` is basically implemented as "go over all the instance fields and see if the marshaller for the field is the one that just copies bytes over". Then there's the logic that chooses the marshallers that is... quite complex.

The question is how much do we need to match this to be compatible. We could store the instance layout information along with our IsBlittable/IsManagedSequential decisions in the image for CoreCLR to deserialize like we discussed before. This would mean we don't need to match CoreCLR exactly.

But if our decisions are different from CoreCLR's, would that be observable to badly annotated p/invokes like dotnet/corefx#39285? (This p/invoke is missing the `OutAttribute` on the pvBuffer parameter, so the runtime wouldn't copy back the values from native, but since CoreCLR considers `USEROBJECTFLAGS` blittable, we just pin the managed value and there's no copy to native/from native. The FX code using the p/invoke considers pvBuffer an out parameter.) Would breaking badly annotated p/invokes be acceptable?

Reply:
For Jan's question regarding the level of testing, within my inner loop I was mostly running the CoreCLR JIT Pri#1 tests in debug / release mode with / without the large version bubble; I however only looked at individual IsManagedSequential calculations manually in the debugger for some types where I originally identified a mismatch between Crossgen and CPAOT. I like the instrumentation idea, that should be relatively straightforward to implement.

I believe we're in general agreement that persisting (some) field information in the R2R file is the way to go but I doubt it's possible to design, implement and sufficiently validate in the planned preview timeframe; so my initial assumption was that, for now, we'd just replicate the CoreCLR logic the best we can, and as a next step we'd proceed towards implementing the persisted field info as a more robust longer-term solution for the problem.